### PR TITLE
[Task#13653] Create Search Screen

### PR DESCRIPTION
**Pull request by cuonghx-0882**
### Issue Link :link:

https://edu-redmine.sun-asterisk.vn/issues/13653
### Goals :soccer:

- Create UI Search Screen
- Handler Event Search, Select Item
- Fetch data and Display

### Implementation Details :construction:

Screen Shot 2019-06-24 at 08 29 20
Screen Shot 2019-06-24 at 08 28 59
Screen Shot 2019-06-24 at 08 28 52
link video: https://www.youtube.com/watch?v=hsBN7X5aTXY

### CODE REVIEW

File: MovieDB/Platform/Repositories/SearchRepository.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        return API.shared
+            .getMovieByKeyword(input: input)
+            .map { output in
+                guard let movies = output.movies else {
+                    throw APIInvalidResponseError()
+                }
+                return PagingInfo<Movie>(page: page,
+                                         items: movies
+                                            .filter { movie in
+                                                if genres.isEmpty {
```
Comment (longvt96) [inline MovieDB/Platform/Repositories/SearchRepository.swift:26]:
```suggestion
                                               return genres.isEmpty ? true : !Set(movie.genres).isDisjoint(with: genres)
```

File: MovieDB/Scenes/Search/SearchUseCase.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    
+    func loadMoreMovie(keyword: String, genres: [Int], page: Int)
+        -> Observable<PagingInfo<Movie>> {
+            if keyword.isEmpty {
+                return movieRepo.getMovieList(category: .popular, page: page).map { pageInfo in
+                    PagingInfo<Movie>(page: page,
+                                      items: pageInfo.items
+                                        .filter { movie -> Bool in
+                                            if genres.isEmpty {
```
Comment (longvt96) [inline MovieDB/Scenes/Search/SearchUseCase.swift:49]:
```suggestion
 return genres.isEmpty ? true : !Set(movie.genres).isDisjoint(with: genres)
```

**Review by longvt96 — APPROVED** [review]

File: MovieDB/Platform/Repositories/SearchRepository.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    func searchMovie(keyword: String, genres: [Int], page: Int) -> Observable<PagingInfo<Movie>> {
+        let input = API.GetMovieByKeywordInput(query: keyword, page: page)
+        return API.shared
+            .getMovieByKeyword(input: input)
+            .map { output in
+                guard let movies = output.movies else {
+                    throw APIInvalidResponseError()
+                }
+                return PagingInfo<Movie>(page: page,
+                                         items: movies
```
Comment (tienpx3010) [inline MovieDB/Platform/Repositories/SearchRepository.swift:24]: let movies = movies.filter

File: MovieDB/Scenes/Search/Cell/GenreCollectionViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import Kingfisher
+
+final class GenreCollectionViewCell: UICollectionViewCell, NibReusable {
+    
+    @IBOutlet weak var nameLabel: UILabel!
+    
+    override var isSelected: Bool {
+        didSet {
+            backgroundColor = isSelected ? UIColor.red : UIColor.groupTableViewBackground
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/Cell/GenreCollectionViewCell.swift:17]:
```suggestion
            backgroundColor = isSelected ? .red : .groupTableViewBackground
```

File: MovieDB/Scenes/Search/Cell/GenreModel.swift [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import RxDataSources
+
+struct GenreModel {
+    var name: String
+    var id: Int
+}
+
+extension GenreModel {
+    init() {
+        self.init(name: "",
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/Cell/GenreModel.swift:18]:
```suggestion
        self.init(name: "", id: 0)
```

File: MovieDB/Scenes/Search/SearchAssembler.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+extension SearchAssembler where Self: DefaultAssembler {
+    func resolve(navigationController: UINavigationController) -> SearchNavigatorType {
+        return SearchNavigator(assembler: self,
+                               navigator: navigationController)
+    }
+    
+    func resolve() -> SearchUseCaseType {
+        return SearchUseCase(searchRepo: resolve(),
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchAssembler.swift:40]:
```suggestion
        return SearchUseCase(searchRepo: resolve(), movieRepo: resolve())
```

File: MovieDB/Scenes/Search/SearchNavigator.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+struct SearchNavigator: SearchNavigatorType {
+    var assembler: Assembler
+    var navigator: UINavigationController
+    
+    func toDetailScreen(movie: Movie) {
+        let detailVC: DetailViewController = assembler.resolve(navigation: navigator,
+                                                               movie: movie)
+        navigator.pushViewController(detailVC,
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchNavigator.swift:22]:
```suggestion
        navigator.pushViewController(detailVC, animated: true)
```

File: MovieDB/Scenes/Search/SearchUseCase.swift [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+//  Copyright © 2019 Sun*. All rights reserved.
+//
+
+import Foundation
+
+protocol SearchUseCaseType {
+    func getListGenres() -> Observable<[GenreModel]>
+    func getGenresID(indexList: [IndexPath]) -> [Int]
+    func searchMovie(keyword: String, genres: [Int]) -> Observable<PagingInfo<Movie>>
+    func loadMoreMovie(keyword: String, genres: [Int], page: Int)
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchUseCase.swift:15]:
```suggestion
    func loadMoreMovie(keyword: String, genres: [Int], page: Int) -> Observable<PagingInfo<Movie>>
```

File: MovieDB/Scenes/Search/SearchViewController.swift [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            }
+            .disposed(by: rx.disposeBag)
+
+        output.selectedMovie
+            .drive()
+            .disposed(by: rx.disposeBag)
+        output.isEmptyData
+            .drive(resultTableView.isEmptyData)
+            .disposed(by: rx.disposeBag)
+        output.gotoTop
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewController.swift:98]: K xử lý trong vm thì drive luôn trong vc

File: MovieDB/Scenes/Search/SearchViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            .asDriverOnErrorJustComplete()
+        
+        let selectedMovie = input.selectionMovie
+            .withLatestFrom(movie) { $1[$0.row] }
+            .do(onNext: { movie in
+                self.navigator.toDetailScreen(movie: movie)
+            })
+        
+        let isEmptyData = checkIfDataIsEmpty(fetchItemsTrigger: fetchItems,
+                                             loadTrigger: Driver.merge(loading,
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewModel.swift:110]:
```suggestion
                                             loadTrigger: Driver.merge(loading, refreshing),
```

File: MovieDB/Scenes/Search/SearchViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+        let searchText = input.textSearch
+            .debounce(0.5)
+            .distinctUntilChanged()
+        
+        let search = searchText
+            .withLatestFrom(input.selectionGenre) { ($0, $1) }
+            .asObservable()
+        
+        let getItems: () -> Observable<PagingInfo<Movie>> = {
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewModel.swift:54]:
```suggestion
        let getItems: (param: (Query, Genres)) -> Observable<PagingInfo<Movie>> = {
```

File: MovieDB/Scenes/Search/SearchViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+                    let (query, selectedItems) = arg
+                    let genres = self.usecase.getGenresID(indexList: selectedItems)
+                    return self.usecase
+                        .loadMoreMovie(keyword: query, genres: genres, page: page)
+                        .trackError(errorTracker)
+                }
+        }
+        
+        let loadTrigger = Driver.merge(input.loadTrigger,
+                                       searchText.mapToVoid(),
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewModel.swift:80]:
```suggestion
                                       searchText,
```

File: MovieDB/Scenes/Search/SearchViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+                                       input.selectionGenre.mapToVoid())
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewModel.swift:81]:
```suggestion
                                       input.selectionGenre)
```

File: MovieDB/Scenes/Search/SearchViewController.swift [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        super.viewDidLoad()
+        configView()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        tabBarController?.title = "Search"
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewController.swift:35]: Bỏ

File: MovieDB/Scenes/Search/SearchViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+            .do(onNext: { movie in
+                self.navigator.toDetailScreen(movie: movie)
+            })
+        
+        let isEmptyData = checkIfDataIsEmpty(fetchItemsTrigger: fetchItems,
+                                             loadTrigger: Driver.merge(loading, refreshing),
+                                             items: movie)
+        let genresList = input.loadTrigger
+            .flatMapLatest {
+                self.usecase
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewModel.swift:115]:
```suggestion
                self.usecase.getListGenres()
```

File: MovieDB/Scenes/Search/SearchViewController.swift [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        output.selectedMovie
+            .drive()
+            .disposed(by: rx.disposeBag)
+        output.isEmptyData
+            .drive(resultTableView.isEmptyData)
+            .disposed(by: rx.disposeBag)
+
+        Driver.combineLatest(selectionGenres,
+                             searchBar.rx.text.orEmpty.asDriver())
+            .map { _ in CGPoint(x: 0, y: 0) }
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewController.swift]: Xử lý khi load lại genresList

File: MovieDB/Scenes/Search/SearchUseCase.swift [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        let (keyword, listSelectedGenres) = param
+        let genres = getGenresID(indexList: listSelectedGenres)
+        if keyword.isEmpty {
+            return movieRepo
+                .getMovieList(category: .popular, page: page)
+                .map { pageInfo in
+                    let movies = pageInfo.items
+                        .filter { movie in
+                            return genres.isEmpty
+                                ? true : !Set(movie.genres).isDisjoint(with: genres)
```
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchUseCase.swift]: Sử dụng ||

File: MovieDB/Scenes/Search/SearchViewController.swift (hunk 16 above)
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewController.swift]: Thêm vào lúc load lại list

File: MovieDB/Scenes/Search/SearchViewController.swift (hunk 14 above)
Comment (tienpx3010) [inline MovieDB/Scenes/Search/SearchViewController.swift:35]: Bỏ

**Review by tienpx3010 — APPROVED** [review]

**Pull request merged**